packages/lib/src/treeUtils/sandbox.ts

```diff
@@ -76,18 +77,22 @@ export class SandboxManager {
       if (this.withSandboxPatchRecorder) {
         throw failure("original subtree must not change while 'withSandbox' executes")
       }
-      this.allowWrite(() => {
-        applyPatches(this.subtreeRootClone, patches)
+
+      this.readonlyMw.writeAllowed = true
+      applyPatches(this.subtreeRootClone, patches)
+
+      // ensure we only stop the write lock until current actions and triggered reactions are finished
+      runAfterCurrentActions(() => {
+        this.readonlyMw.writeAllowed = false
```

Comment:
Would it be better/safer to do the following similar to `allowWrite`?

```ts
const writeAllowedOld = this.readonlyMw.writeAllowed
this.readonlyMw.writeAllowed = true
applyPatches(this.subtreeRootClone, patches)

// ensure we only stop the write lock until current actions and triggered reactions are finished
runAfterCurrentActions(() => {
  this.readonlyMw.writeAllowed = writeAllowedOld
})
```

Reply:
Thanks, fixed it in a even more secure manner.

I still feel torn about this though since in this case the exception that readonly middleware throws actually saves you from a reaction that makes the sandbox and the original node get out of sync :-/

Reply:
> Thanks, fixed it in a even more secure manner.

Looks even better! :-)

> I still feel torn about this though since in this case the exception that readonly middleware throws actually saves you from a reaction that makes the sandbox and the original node get out of sync :-/

I know what you mean ... In case you didn't see it yet, [@mweststrate doesn't even seem to like the whole sandbox idea](https://github.com/mobxjs/mobx/issues/2259#issuecomment-576199561), but I still believe that it's the only high-performance way to test "what-if" scenarios. I think the costly part is tweaking nodes during `fromSnapshot`, perhaps setting up observers etc., so the more robust `clone` every time a copy is needed, just won't work if this needs to be done at several places for each animation frame.

Reply:
@mweststrate's comment about not using reactions to update state

> 1. As documented, updating state from reactions is generally an anti pattern, probably you need computeds, which wouldn't suffer from this idea

got me thinking and perhaps I actually need to change my state design to use a computed property where I'm currently using a reaction.

I'd say, don't merge this PR for now because I share your skepticism about this solution and maybe it's my "un-MobX" way of doing things with a reaction in this case, which shouldn't affect `mobx-keystone`.